api/renter.go

```diff
@@ -350,49 +387,141 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
 
 // renterDownloadHandler handles the API call to download a file.
 func (api *API) renterDownloadHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	p, errmsg := api.parseAndValidateDownloadParameters(w, req, ps)
+	if errmsg != nil {
+		WriteError(w, *errmsg, http.StatusBadRequest)
 		return
 	}
 
-	err := api.renter.Download(strings.TrimPrefix(ps.ByName("siapath"), "/"), destination)
-	if err != nil {
-		WriteError(w, Error{"download failed: " + err.Error()}, http.StatusInternalServerError)
-		return
+	if p.Async { // Create goroutine if `async` param set.
```

Comment:
I think that the input validation should probably be done alongside the use. When you validate separately, the likelihood of missing something goes up, especially if in the future this part of the code is modified. The requirements may change, and then the writer and reviewer may forget to update the validator. But if the validation happens alongside the use, then during review it's a lot easier to realize that you've written code which invalidates your validation.